### PR TITLE
csi: plugins track jobs in addition to allocations, and use job information to set expected counts

### LINKED ISSUE #8503: CSI plugin expected counts reflect expected instance counts more accurately
CSI plugin `Expected` counts are currently based only on the plugin fingerprint data from clients, including both healthy and unhealthy fingerprints. The count would be more useful for operators if it reflected the expectation from the jobs

- for service jobs (i.e. controllers) the `count` from the job configuration
- for system jobs (i.e. nodes) the number of running allocations + blocked evaluations

### PR DESCRIPTION
**Pull request by langmartin**
Expected counts are derived from jobs, we may expect plugins for which
we have no valid fingerprints and therefore no allocations.

- On job update, update plugin job collection and re-count expected

- System jobs expected count is the sum of currently running
  allocations + blocked evals. Blocked evals number will improve in
  accuracy with blocking that accounts for driver start time.
  It is called in `updateJobSummaryByAllocation`
  
  This does require keeping the expected count indexed by jobID so
  that we can update on each allocation change.

- Plugin emptiness accounts for jobs

Closes #8503 See also #7974

### CODE REVIEW

File: nomad/state/state_store_test.go [hunk 1]
```diff
@@ -3522,8 +3522,8 @@ func TestStateStore_CSIPluginMultiNodeUpdates(t *testing.T) {
 
 func TestStateStore_CSIPluginJobs(t *testing.T) {
 	s := testStateStore(t)
-	deleteNodes := CreateTestCSIPlugin(s, "foo")
-	defer deleteNodes()
+	// deleteNodes := CreateTestCSIPlugin(s, "foo")
+	// defer deleteNodes()
```
Comment (tgross) [inline nomad/state/state_store_test.go:3526]: Was this just left behind from testing?

File: nomad/state/state_store.go [hunk 2]
```diff
@@ -4498,6 +4514,70 @@ func (s *StateStore) updateJobScalingPolicies(index uint64, job *structs.Job, tx
...
+			for _, t := range tg.Tasks {
+				if t.CSIPluginConfig == nil {
+					continue
+				}
+
+				plugIn, ok := plugIns[t.CSIPluginConfig.ID]
+				if !ok {
+					p, err := s.CSIPluginByID(ws, t.CSIPluginConfig.ID)
+					if err != nil {
+						return fmt.Errorf("%v", err)
```
Comment (tgross) [inline nomad/state/state_store.go:4533]: We can probably just `return err` here and elsewhere in this `updateJobCSIPlugins` method where we're not adding any new context to the error.

File: nomad/state/state_store.go [hunk 3]
```diff
@@ -4498,6 +4514,70 @@ func (s *StateStore) updateJobScalingPolicies(index uint64, job *structs.Job, tx
 	return nil
 }
 
+// updateJobCSIPlugins runs on job update indexes the job in the plugin
```
Comment (tgross) [inline nomad/state/state_store.go:4517]:
```suggestion
// updateJobCSIPlugins runs on job update, and indexes the job in the plugin
```

File: nomad/structs/csi.go [hunk 4]
```diff
@@ -874,6 +890,118 @@ func (p *CSIPlugin) DeleteAlloc(allocID, nodeID string) error {
...
+
+type JobDescription struct {
+	Namespace string
+	ID        string
+	Version   uint64
+	Expected  int
+}
+
+type JobNamespacedDescriptions map[string]JobDescription
+type JobDescriptions map[string]JobNamespacedDescriptions
```
Comment (tgross) [inline nomad/structs/csi.go:960]: Can we annotate the key for these maps? You can figure it out from context but having a comment here makes it easier to know what that `string` is at a glance.

File: nomad/state/state_store.go [hunk 5]
```diff
@@ -4759,6 +4842,41 @@ func (s *StateStore) updatePluginWithAlloc(index uint64, alloc *structs.Allocati
...
+
+	for _, t := range tg.Tasks {
+		if t.CSIPluginConfig != nil {
+			pluginID := t.CSIPluginConfig.ID
+			plug, err := s.CSIPluginByID(ws, pluginID)
+			if err != nil {
+				return err
+			}
+			if plug == nil {
+				plug = structs.NewCSIPlugin(pluginID, index)
```
Comment (tgross) [inline nomad/state/state_store.go:4864]: If understand this and the changes in `upsertNodeCSIPlugins` correctly, we'll now see the plugin registered before the fingerprint. (Which I think is a good UX change, in addition to everything else!) It looks like the scheduler should still block placements until the fingerprint is done, but is that only because we don't associate the allocations with the plugin here?

File: nomad/state/state_store_test.go (hunk 1 above)
Comment (langmartin) [reply nomad/state/state_store_test.go:3526]: yes, that was in place before jobs created plugins successfully

File: nomad/state/state_store.go (hunk 2 above)
Comment (langmartin) [reply nomad/state/state_store.go:4533]: that's my emacs snippet 🤦

File: nomad/state/state_store.go (hunk 5 above)
Comment (langmartin) [reply nomad/state/state_store.go:4864]: Actually, if we switched to recording the allocations here and just setting their fingerprint dependent properties on node update, the scheduler would still block on `ControllersHealthy > 0`. I don't think going that far would be a UX visible change though.

File: nomad/structs/csi.go (hunk 4 above)
Comment (langmartin) [reply nomad/structs/csi.go:960]: 👍

**Comment by langmartin** [comment]
Well, we'd definitely benefit from the bigger test! This change introduces an edge case where a plugin has been created by a job and that job is deleted before any allocation fingerprints make it back to the state store. The plugin will exist, but not have any allocations to use for GC. Working on a fix now.

File: nomad/structs/csi.go [hunk 6]
```diff
@@ -874,6 +890,128 @@ func (p *CSIPlugin) DeleteAlloc(allocID, nodeID string) error {
...
+	if j == nil {
+		j = make(JobDescriptions)
+	}
+	if j[job.Namespace] == nil {
+		j[job.Namespace] = make(JobNamespacedDescriptions)
+	}
+	j[job.Namespace][job.ID] = JobDescription{
+		Namespace: job.Namespace,
+		ID:        job.ID,
+		Version:   job.Version,
```
Comment (tgross) [inline nomad/structs/csi.go:979]: If we're using the version as a value and not part of the key, that implies to me that the later version of a job is intended to overwrite the values for a newer version. Can you walk me through what we expect to see here in the case of a job update with canaries or a rollback?

File: nomad/csi_endpoint_test.go [hunk 7]
```diff
@@ -689,6 +689,83 @@ func TestCSIPluginEndpoint_RegisterViaFingerprint(t *testing.T) {
 	require.Nil(t, resp2.Plugin)
 }
 
+func TestCSIPluginEndpoint_RegisterViaJob(t *testing.T) {
```
Comment (tgross) [inline nomad/csi_endpoint_test.go:692]: Solid new test 👍

File: nomad/state/state_store.go (hunk 5 above)
Comment (tgross) [reply nomad/state/state_store.go:4887]: Ok that doesn't seem like something we want to do in this pass. Let's keep what you've got here.

File: nomad/structs/csi.go (hunk 6 above)
Comment (langmartin) [reply nomad/structs/csi.go]: I've talked to @DingoEatingFuzz, and the API changes are not necessary; the UI has to make the same number of round trips to get the required data. I think for now the best idea is just to back out the API change, which I've done.

**Comment by langmartin** [comment]
> Assuming we've done some end-to-end testing of this, it looks ok to me. I've left a question about the expected behavior around updates.

End to end is looking good.
